File: summarize.py
#!/usr/bin/python

import csv
import glob
import json
import os, sys
import datetime

os.chdir(sys.path[0])    # cd to the script dir, even if called from cron

for logfile in glob.glob('logs/*.log'):
    visitorsfile = logfile[:-4] + '.visitors'
    tmpfile = logfile + '.tmp'
    referersfile = logfile[:-4] + '.referers'

    S = dict()

    try:
        with open(visitorsfile, 'r') as infile:
            visitors = json.load(infile)
    except:
        visitors = dict()

    try:
        with open(referersfile, 'r') as infile:
            referers = set(json.load(infile))
    except:
        referers = set()

    today = datetime.datetime.today().strftime('%Y-%m-%d')

    with open(logfile, 'rb') as inp, open(tmpfile, 'w') as out:
        writer = csv.writer(out, delimiter='\t')
        for row in csv.reader(inp, delimiter='\t'):
            rowdate = datetime.datetime.fromtimestamp(int(row[0])).strftime('%Y-%m-%d')
            ip = row[1]
            ua = row[3]            
            
            if any(excl in ua for excl in ['bot', 'crawl', 'slurp', 'spider', 'yandex']):                
                continue            

            referer = row[4]
            if referer != "":
                referers.add(referer)

            if rowdate not in S:        
                S[rowdate] = set()      # let's use a set in order to count each IP only once a day

            S[rowdate].add(ip)
            if rowdate == today:
                writer.writerow(row)

    os.remove(logfile)
    os.rename(tmpfile, logfile)

    for k in S:
        visitors[str(k)] = str(len(S[k]))

    with open(visitorsfile, 'w') as outfile:
        json.dump(visitors, outfile)

    with open(referersfile, 'w') as outfile:
        json.dump(sorted(list(referers)), outfile)

    with open('.lastsummarize', 'w') as outfile:
        outfile.write(str(int(time.time())))

    os.chmod(logfile, 0o666)
    os.chmod(visitorsfile, 0o666)
    os.chmod(referersfile, 0o666)

File: summarize.php
<?php
// Get a list of sites
$logfiles = glob('logs/*.log', GLOB_BRACE);
// Iterate through each site's logs
foreach($logfiles as $logfile) {
  // Name our working files based on the main site log name
  $visitorsfile = substr($logfile,0,-4).'.visitors';
  $referersfile = substr($logfile,0,-4).'.referers';
  $tmpfile = $logfile.'.tmp';
  
  // Check we can open the log file, then process it 
  if (($log = fopen($logfile, "r")) !== FALSE) {

    // While we have rows available in the log file, do stuff
    while (($data = fgetcsv($log, 1000, "\t")) !== FALSE) {

      // Assign variables to the data
      $rowdate = date("Y-m-d", $data[0]);
      $ipadr = $data[1];
      $ua = $data[3];
      $referer = $data[4];

      // If the User Agent is from a bot, then skip it
      $bots = array('bot', 'crawl', 'slurp', 'spider', 'yandex', 'WordPress', 'AHC', 'jetmon');
      foreach($bots as $bot){
        if (strpos($ua, $bot) !== false) {
          continue;
        }
      }
      
      // We only count each IP once a day (Unique daily visitors)
      if ( in_array($ipadr,$ipTracker[$rowdate]) ) {
        continue; 
      }
      
      // We only get this far for unique, non-bot data, so let's start recording it
      $visitors[$rowdate]++;
      
      // If we have referer data, add it to the array (if it doesn't yet exist)
      if ( $referer != "" && !in_array($referer,$referers)) {
        $referers[] = $referer;
      }

      // Track the IP per date
      $ipTracker[$rowdate][] = $ipadr;
      
    }
    fclose($log);
    
    // Write the Visitors file
    $writefile = fopen($visitorsfile,"w");
    fwrite($writefile,json_encode($visitors));
    fclose($writefile);
    
    // Write the Referrers file
    $writefile = fopen($referersfile,"w");
    fwrite($writefile,json_encode($referers));
    fclose($writefile);
    
    // Clear our variables for the next site
    unset($visitors);
    unset($referers);
    unset($ipTracker);
  }
}
// Update the .lastsummarize
$writefile = fopen('.lastsummarize',"w");
fwrite($writefile,time());
fclose($writefile);
